framework/src/main/java/org/checkerframework/framework/util/dependenttypes/DependentTypesHelper.java

```diff
@@ -592,195 +497,285 @@ public void standardizeVariable(AnnotatedTypeMirror type, Element elt) {
+     * parameter), the expression is removed from the annotation. This could result in dependent
+     * type annotations with empty lists of expressions. If this is a problem, subclasses can
+     * override {@link #buildAnnotation(AnnotationMirror, Map)} to do something besides creating an
+     * annotation with a empty list.
      *
-     * @param localVarPath the expression is parsed as if it were written at this location
-     * @param type the type to parse; is side-effected by this method
+     * @param methodDeclTree the method declaration to which the annotations are viewpoint-adapted
+     * @param atm type to viewpoint-adapt; is side-effected by this method
      */
-    private void parseToPath(TreePath localVarPath, AnnotatedTypeMirror type) {
-        Tree enclosingClass = TreePathUtil.enclosingClass(localVarPath);
-        TypeMirror enclosingType = TreeUtils.typeOf(enclosingClass);
-
-        JavaExpression receiver = JavaExpression.getPseudoReceiver(localVarPath, enclosingType);
-
-        JavaExpressionContext localContext =
-                new JavaExpressionContext(
-                        receiver,
-                        JavaExpression.getParametersOfEnclosingMethod(localVarPath),
-                        factory.getChecker());
-        standardizeAtm(localContext, localVarPath, type, /*removeErroneousExpressions=*/ false);
+    public void delocalize(MethodTree methodDeclTree, AnnotatedTypeMirror atm) {
+        if (!hasDependentType(atm)) {
+            return;
+        }
+
+        TreePath pathToMethodDecl = factory.getPath(methodDeclTree);
+
+        ExecutableElement methodElement = TreeUtils.elementFromDeclaration(methodDeclTree);
+        StringToJavaExpression stringToJavaExpr =
+                expression -> {
+                    JavaExpression result;
+                    try {
+                        result =
+                                StringToJavaExpression.atPath(
+                                        expression, pathToMethodDecl, factory.getChecker());
+                    } catch (JavaExpressionParseException ex) {
+                        return null;
+                    }
+                    List<FormalParameter> parameters =
+                            JavaExpression.getFormalParameters(methodElement);
+                    List<JavaExpression> paramsAsLocals =
+                            JavaExpression.getParametersAsLocalVariables(methodElement);
+                    JavaExpressionConverter jec =
+                            new JavaExpressionConverter() {
+                                @Override
+                                protected JavaExpression visitLocalVariable(
+                                        LocalVariable localVarExpr, Void unused) {
+                                    int index = paramsAsLocals.indexOf(localVarExpr);
+                                    if (index == -1) {
```

Comment:
This method could throw an exception that is caught at the call to `jec.convert`.
That has the small advantages of not having to create a new object which will never be used for anything (other than being tested against), and if there is a problem it is signalled right away rather than continuing to visit the expression.

If you think that is too ugly, we can keep the code as is.